keps/sig-api-machinery/storage-migration-auto-trigger.md

```diff
@@ -0,0 +1,229 @@
+}
+
+type StorageStateSpec {
+  // The resource this StorageState is about.
+  Resource GroupResource
+}
+
+type StorageStateStatus {
+  // The hash values of possible stored versions, the versions to which the
+  // persisted instances of spec.resource are encoded.
+  // "Unknown" is a valid value in the list, and is the default value.
+  // It is not safe to upgrade or downgrade to an apiserver binary that does not
+  // support all versions listed in this field, or if "Unknown" is listed.
+  // Once the storage version migration for this resource has completed, the
+  // value of this field is refined to only contain the storageVersionHash.
+  // Once the apiserver has changed the storage version, the new storage version
+  // is appended to the list.
+  // +optional
+  StoredVersionsHashes []string
+  // The hash value of the storage version, as shown in the discovery document
+  // served by the API server.
+  // Storage Version is the version to which objects are converted to
+  // before persisted.
+  StorageVersionHash string
+  // LastHeartbeatTime is the last time the storage migration triggering
+  // controller checks the storage version hash of this resource in the
+  // discovery document and updates this field.
+  // +optional
+  LastHeartbeatTime metav1.Time
+}
+```
+
+We had considered making `StoredVersionsHashes` part of the `spec` instead of
+the `status`, because the stored versions cannot be deduced immediately by
+inspecting Kubernetes API. However, we decided to put it in the status because
+* Its value eventually is determined by the `StorageVersionMigration` API, i.e.,
+  it's set to the storage version when the corresponding migration has
+  completed.
+* Even though we cannot reconstruct this field immediately, this field is
+  initialized to empty, meaning "unknown", which truthfully reflects the state.
+* Putting it in the spec is even more unnatural. Spec describes the desired
+  state the system should eventually reach in the future, while this field is
+  about states in the history.
+* Keeping it in the status is consistent with the [CRD.status.StoredVersions][]
+  API.
+
+[CRD.status.StoredVersions]:https://github.com/kubernetes/kubernetes/blob/697c2316faaabae8ef8371032b60be65d7795e68/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go#L305
+  
+## Storage migration triggering controller
+
+We will add a controller that monitors the discovery documents and the
+storageStates to trigger storage version migration.
+
+The controller fetches the discovery document periodically. For each resource,
+if the storageState doesn't exist for the resource, the controller
+  * stops any ongoing migration of this resource by deleting existing
+    [storageVersionMigrations][],
+  * creates a new storageVersionMigration for this resource,
+  * creates the storageState for the resource, setting
+    .status.storedVersionsHash to ["Unknown"], and status.StorageVersionHash to
```

Comment:
s/.status.storedVersionsHash/.status.storedVersionsHashes/

Reply:
do you want just [] or ["Unknown"]?

Reply:
Explicitly lists ["Unknown"] is easier to manage. I'll stick to it.

Reply:
ok, you still need s/.status.storedVersionsHash/.status.storedVersionsHashes/